### PR TITLE
Refactor Integer Generator

### PR DESCRIPTION
**Pull request by gogit**
Refactor Integer Generator

### CODE REVIEW

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java [hunk 1]
```diff
@@ -1,15 +1,64 @@
...
+     * Package Access only
+     * 
+     * @param min value included
+     * @param max value excluded
+     * @see RandomGenerator
+     */
+    IntegerGenerator(final int min, final int max) {
+        if (min >= max) {
+            throw new IllegalArgumentException(
+                            format("%s cannot be greater or equal to %s", min, max));
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java]: Could have better error message? 
E.g.
Min value cannot be greater than or equal to Max value, got Min: %s and Max: %s

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java [hunk 2]
```diff
@@ -1,15 +1,64 @@
...
         this.max = max;
     }
 
     @Override
     public Integer next() {
-        return RANDOM.nextInt(max);
+        if(min >= 0){
+            return ThreadLocalRandom.current().nextInt(min, max);
+        }
+        long posMin = -(long)min;
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java]: Additional variable posMin not required. Same result can be achieved without that.

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java [hunk 3]
```diff
@@ -1,15 +1,64 @@
...
+        long range = (long)max + posMin;
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java]: Need final on the variable

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java [hunk 4]
```diff
@@ -1,15 +1,64 @@
...
+        }
+        this.min = min;
         this.max = max;
     }
 
     @Override
     public Integer next() {
-        return RANDOM.nextInt(max);
+        if(min >= 0){
+            return ThreadLocalRandom.current().nextInt(min, max);
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java]: Any benefit of using ThreadLocalRandom.current().nextInt(min, max) over RANDOM.ints(min, max).limit(1).findFirst().getAsInt(). The reason I am mentioning is we already have RANDOM as part of the parent class and it's easy to mock the RANDOM in testing, if required.

Comment (gogit) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/IntegerGenerator.java]: true its a round about way but has its merits for mocking

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    public void shouldReturnIntegersBetweenMinimumInclusiveAndMaximumExcluded() {
+        IntegerGenerator ig = new IntegerGenerator();
+        for (int i = 0; i < 10000; i++) {
+            int k = ig.next();
+            assertTrue(k < Integer.MAX_VALUE);
+            assertTrue(k >= Integer.MIN_VALUE);
+        }
+    }
+       
+    private void shouldTestRangesUsingMinMaxGenerator(final int min, final int max, final int loopCount) {
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: Same as below

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java [hunk 6]
```diff
@@ -0,0 +1,218 @@
+package uk.gov.justice.services.test.utils.core.random;
+
+import static org.junit.Assert.assertTrue;
+
+import org.junit.Test;
+
+public class IntegerGeneratorTest {
```
Comment (deepakhalale) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: Not sure if we are adding any value by having lots of test cases which are only testing different ranges? I will leave someone else (Allan or Matt) to decide on this.

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java (hunk 5 above)
Comment (gogit) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: after we refactor it to make it more readable

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java (hunk 6 above)
Comment (gogit) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: if you are generating random numbers, range checking ensures that it works in those ranges and not just in a blanket range.

Comment (deepakhalale) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: Need a test case which tests using Integer.MIN_VALUE and Integer.MAX_VALUE. This is to check that code handles the overflow case.

Comment (gogit) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/IntegerGeneratorTest.java]: Does not apply in this generator

**Pull request merged**